network/validator/pubsub/authorized_sender_validator_test.go

```diff
@@ -0,0 +1,120 @@
+package validator
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/model/messages"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/network"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type TestCase struct {
+	Identity   *flow.Identity
+	Channel    network.Channel
+	Message    interface{}
+	MessageStr string
+}
+
+func TestIsAuthorizedSender(t *testing.T) {
+	suite.Run(t, new(TestIsAuthorizedSenderSuite))
+}
+
+type TestIsAuthorizedSenderSuite struct {
+	suite.Suite
+	happyPathTestCases []TestCase
+	sadPathTestCases   []TestCase
+}
+
+func (s *TestIsAuthorizedSenderSuite) SetupTest() {
+	s.initializeTestCases()
+}
+
+// TestIsAuthorizedSender_AuthorizedSender checks that IsAuthorizedSender does not return false positive
+// validation errors for all possible valid combinations (authorized sender role, message type).
+func (s *TestIsAuthorizedSenderSuite) TestIsAuthorizedSender_AuthorizedSender() {
+	for _, c := range s.happyPathTestCases {
+		str := fmt.Sprintf("role (%s) should be authorized to send message type (%s) on channel (%s)", c.Identity.Role, c.MessageStr, c.Channel)
+		s.Run(str, func() {
+			msgType, err := IsAuthorizedSender(c.Identity, c.Channel, c.Message)
+			s.Require().NoError(err)
+			s.Require().Equal(c.MessageStr, msgType)
+		})
+	}
+}
+
+// TestIsAuthorizedSender_UnAuthorizedSender checks that IsAuthorizedSender return's ErrUnauthorizedSender
+// validation error for all possible invalid combinations (unauthorized sender role, message type).
+func (s *TestIsAuthorizedSenderSuite) TestIsAuthorizedSender_UnAuthorizedSender() {
+	for _, c := range s.sadPathTestCases {
+		str := fmt.Sprintf("role (%s) should not be authorized to send message type (%s) on channel (%s)", c.Identity.Role, c.MessageStr, c.Channel)
+		s.Run(str, func() {
+			msgType, err := IsAuthorizedSender(c.Identity, c.Channel, c.Message)
+			s.Require().ErrorIs(err, ErrUnauthorizedSender)
+			s.Require().Equal(c.MessageStr, msgType)
+		})
+	}
+}
+
+// TestIsAuthorizedSender_ValidationFailure checks that IsAuthorizedSender returns the expected validation error.
+func (s *TestIsAuthorizedSenderSuite) TestIsAuthorizedSender_ValidationFailure() {
+	s.Run("sender is ejected", func() {
+		identity := unittest.IdentityFixture()
+		identity.Ejected = true
+		msgType, err := IsAuthorizedSender(identity, network.Channel(""), nil)
+		s.Require().ErrorIs(err, ErrSenderEjected)
+		s.Require().Equal("", msgType)
+	})
+
+	s.Run("unknown message type", func() {
+		identity := unittest.IdentityFixture()
+		msgType, err := IsAuthorizedSender(identity, network.Channel(""), nil)
```

Comment:
Maybe constructing an `unknownMessageType` for testing and passing it instead of `nil`. Also, good to keep both `nil` and `unknownMessageType` inputs together.

Reply:
https://github.com/onflow/flow-go/pull/2648/commits/f848585d4d1623be7953c851c92ad7fb53e19909